### ISSUE TITLE
Some issue with going to page via link vs just reloading page with url. Via link renders nothing.

### ISSUE THREAD

**Issue by idchlife** [developit/preact-router#56]
Here is the video of behaviour. Keep in mind, when I remove modifiers, behaviour is normal.

https://monosnap.com/file/6XgVsdl9lfhBuSi1gcCL1MMSvggzil

Here you can see that it does not even render anything:

![image](https://cloud.githubusercontent.com/assets/4563032/17170091/4ac91caa-53f3-11e6-8b7c-4e758ca07f8b.png)

It should be strongly related to the case that I'm creating 'Modifiers' for my components, that does modify their behaviour. Their code and idea is shown here: http://www.webpackbin.com/VkXuBY_wb

I tried to recreate errors and empty renders but could not.

TL;DR; about modifiers: in constructor they replace cWM, cWU, render methods with their own, checking and doing something before calling to original methods of instance.

preact-router does some caching of components and cannot find one with modified behaviour?

Also, modifiers have redirects in them also, in my case for Authentication state purpose.

**Comment by developit** [comment]
Are the redirects calling `route()` synchronously? There is currently an issue with `preact-router` that causes this type of rendering bug if `route()` is called during a render.

**Comment by idchlife** [comment]
@developit Yes, yo told me this quick fix before. The main thing I think in this parts:

![image](https://cloud.githubusercontent.com/assets/4563032/17173553/27a6fd8a-5404-11e6-8b89-44bfefd29e39.png)

The main 'problem'. When I comment those returns 'something in progress...' everything seems to be fine. Of course, if checking for something will be longer than 1 second, user will have blank screen.

**Comment by developit** [comment]
Couple more questions to help me diagnose:
Is this with Preact 5.6 and preact-router 2.1?  Also, the return value from `oldRender()` - is there a different element type being returned at the root (ie, not a `<div>`)?
Thanks :)

**Comment by idchlife** [comment]
@developit nope, preact 5.4 and preact-router 2.0

Just updated to versions you told.

Now I have everywhere this before everything:

![image](https://cloud.githubusercontent.com/assets/4563032/17177487/4e5f9660-541a-11e6-94f4-2c32817826f5.png)

It just being prepended to the other render.

![image](https://cloud.githubusercontent.com/assets/4563032/17177554/9fda430a-541a-11e6-967b-f926cb2b1aa9.png)

**Comment by developit** [comment]
@idchlife interesting - what is the root of your call to `preact.render()` like - are you rendering a `<Router />` at the root? (or a tree of components that ends up rendering a router at the root?)

(something like this might be informative:)

```js
console.log(
  [].map.call(document.body.childNodes, c => c._component.constructor.name)
);
```

**Comment by idchlife** [comment]
@developit 

![image](https://cloud.githubusercontent.com/assets/4563032/17212564/b5b6cc0c-54d8-11e6-9354-4bfc3b914fdb.png)

![image](https://cloud.githubusercontent.com/assets/4563032/17212600/de0d8e70-54d8-11e6-9e7f-28562f5854c9.png)


Is this helpfull?

**Comment by developit** [comment]
Yes! We're getting closer. It looks like an issue with re-rendering the root. If you wrap your render in a jsx `<div>` it should work.

**Comment by developit** [comment]
@idchlife I think I have a repro for this issue, will try to publish a fix this weekend!

**Comment by mdierolf** [comment]
I am experiencing this issue as well. Wrapping with a `<div>` did not work around the issue for me, but this did:

```
let Wrap = ({ children, ...props }) => {
 children.map( (child) => child.attributes = props);
 return <div>{ children }</div>;
};
...
<Router>
    <Wrap path="/"><HomePage /></Wrap>
</Router>

```

Seems like the problem does not appear when using functional components?

**Comment by developit** [comment]
@mdierolf It doesn't work when you do `<Router><HomePage path="/" /></Router>` ?  It'd be good to know which version of Preact you're seeing that with, since there is a bug in 5.7 that could cause high order components to be incorrectly removed when a high order parent is mutated - that bug might produce a similar looking result to this one.  One way to check if it's that would be to `npm i preact@5.5`. The bug is fixed on master and will be released likely today.

FWIW there is actually a `Route` component exported by `preact-router` that looks a lot like your `Wrap`:

```js
import { Router, Route } from 'preact-router';
<Router>
  <Route path="/">
    <HomePage />
  </Route>
</Router>
```

**Comment by mdierolf** [comment]
@developit it **usually** works with `<Router><HomePage path="/" /></Router>` but frequently fails when using the back/forward buttons, and occasionally fails when using the <Link> component.

I don't have a good repeatable test case otherwise I would send that instead, just wanted to show the workaround in case that was helpful.

I am currently using 5.6.0 - I will give 6.0 a try

**Comment by mdierolf** [comment]
@devlopit after switching to preact 6.0 the problem disappears - thank you!

**Comment by developit** [comment]
@mdierolf heyo, you made my day!

**Comment by developit** [comment]
Now we just need to know if @idchlife's original all issue is also resolved!

**Comment by rrushton** [comment]
Just thought I would say it fixed my issue as well. I was trying the demo code from preact-router and when I clicked the Profile, John or Error items then back to home, it would stop working. Updating to 6.0 fixed it.

**Comment by developit** [comment]
@rrushton Great!  Yes, I've just published an update to [preact-boilerplate](https://github.com/developit/preact-boilerplate/releases/tag/5.0.0) that updates to preact 6 and removes some things people weren't really using. If you are curious, you can see the diff [here](https://github.com/developit/preact-boilerplate/commit/3f9d68343c4300e755dfe9627ba8e82815ac5cf1) 🎉

**Comment by developit** [comment]
I think since we've got 2 confirmations that this is fixed via Preact 6 we can close the issue. Hopefully it worked for @idchlife too!!
